### PR TITLE
[ISSUE-455] Lazily create uncompressedData

### PR DESCRIPTION
**Pull request by xianjingfeng**
### What changes were proposed in this pull request?

Lazily create `uncompressedData`.

### Why are the changes needed?

Save memory. #455 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

The existing UTs is enough

### CODE REVIEW

File: client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java [hunk 1]
```diff
@@ -69,13 +70,7 @@ public RssShuffleDataIterator(
     this.shuffleReadClient = shuffleReadClient;
     this.shuffleReadMetrics = shuffleReadMetrics;
     this.codec = Codec.newInstance(rssConf);
-    // todo: support off-heap bytebuffer
-    this.uncompressedData = ByteBuffer.allocate(
-        (int) rssConf.getSizeAsBytes(
-            RssClientConfig.RSS_WRITER_BUFFER_SIZE,
```
[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:75]: Yes, we should remove this. The concept of RSS_WRITER_BUFFER_SIZE is the threshold of writing buffer, which is not related with reading buffer.

In this PR, we just need to remove this part logic. WDYT? @xianjingfeng

[the remark this answers is not in the document]
Comment (xianjingfeng) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:75]: Done

**Review by jerqi — DISMISSED** [review]
LGTM

File: client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java [hunk 2]
```diff
@@ -69,13 +67,6 @@ public RssShuffleDataIterator(
     this.shuffleReadClient = shuffleReadClient;
     this.shuffleReadMetrics = shuffleReadMetrics;
     this.codec = Codec.newInstance(rssConf);
-    // todo: support off-heap bytebuffer
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java:72]: We would better not remove this `todo` comment.

File: client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java [hunk 3]
```diff
@@ -120,10 +115,20 @@ public boolean hasNext() {
         shuffleReadMetrics.incRemoteBytesRead(compressedDataLength);
 
         int uncompressedLen = compressedBlock.getUncompressLength();
-        if (uncompressedData == null || uncompressedData.capacity() < uncompressedLen) {
+        if (uncompressedData == null) {
+          // todo: support off-heap bytebuffer
+          uncompressedData = ByteBuffer.allocate(
+              (int) rssConf.getSizeAsBytes(
+                  RssClientConfig.RSS_WRITER_BUFFER_SIZE,
```
[the remark this answers is not in the document]
Comment (zuston) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/reader/RssShuffleDataIterator.java]: Yes. The initial `uncompressedData` size is hard to set. I prefer to remove it.

**Review by zuston — DISMISSED** [review]
LGTM

**Review by jerqi — APPROVED** [review]

**Review by zuston — APPROVED** [review]

**Comment by zuston** [comment]
Merged. Thanks @jerqi @xianjingfeng @advancedxy

**Pull request merged**